OpenTrons/opentrons-api
# refactor(api,robot-server): Expose protocol metadata to HTTP clients

# Overview

This PR exposes protocol metadata (like author name and protocol description) through the HTTP API. Closes #8003.

# Code review requests

* Anyone have a better term for this than "pre-analysis"?
* Check out the `protocol_models.Metadata` docstring. Do we agree on this rough definition of metadata, from the perspective of integrators and protocol authors? Further specification is ticketed as #8334.
* Are we still sure about committing to metadata always being available immediately at upload time? Alternatively, the API for reading metadata could work the same way as our API for getting required equipment: a background task could start in the background when you upload the protocol, and you could fetch its results when it finishes, later. (Which might be in 0.1 seconds, or it might be in 30 seconds.)

# Manual test plan

I've done this:

1. Enable the **Enable experimental Protocol Engine** feature flag.
2. `POST /protocols` with a JSON or Python protocol. @mcous's [Postman HTTP collection](https://gist.github.com/mcous/7dde93b6ea83753b86b0d84b40ba07f6) can help.
3. Confirm that the protocol's metadata appears in the response.
    * All of the protocol's metadata fields should be passed straight through. Their types should be retained.
    * There should be no fields in the response that the protocol did *not* explicitly set.
4. Try some error cases:
    * Multi-file protocols should currently raise a `NotImplementedError`
    * Other problems with parsing or analysis should respond with an HTTP 400

This should work with `make dev` as long as you have the right feature flags set; you don't need a real robot.

# Risk assessment

Low. This only touches stuff in Protocol Engine and the new HTTP API.
